pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,396 @@
+	}
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Store current send info
+	m.latencyStore.UpdateNodeIPLatencyEntryByKey(addr.String(), &NodeIPLatencyEntry{
+		SeqID:           seqID,
+		LastSendTime:    timeStart,
+		LastRecvTime:    time.Time{},
+		LastMeasuredRTT: 0,
+	})
+
+	return nil
+}
+
+func (m *MonitorTool) recvPing(socket net.PacketConn, isIPv4 bool, stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			readBuffer := make([]byte, 1024)
+			_, peer, err := socket.ReadFrom(readBuffer)
+			if err != nil {
+				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+					klog.ErrorS(err, "Timeout reading ICMP message")
+					continue
+				}
+				klog.ErrorS(err, "Failed to read ICMP message")
+			}
+
+			destIP := peer.String()
+			// Get the node name by destIP
+			entry, ok := m.latencyStore.GetNodeIPLatencyEntryByKey(destIP)
+			if !ok {
+				klog.ErrorS(err, "Failed to get node name by destIP")
+				continue
+			}
+
+			// Parse the ICMP message
+			var msg *icmp.Message
+			if isIPv4 {
+				msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer)
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+			} else {
+				msg, err = icmp.ParseMessage(ProtocolIPv6ICMP, readBuffer)
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+			}
+
+			echo, ok := msg.Body.(*icmp.Echo)
+			if !ok {
+				klog.ErrorS(nil, "Failed to assert type as *icmp.Echo")
+				continue
+			}
+
+			// Parse the ICMP data
+			if entry.SeqID != uint32(echo.Seq) {
+				klog.ErrorS(err, "Failed to match seqID")
+				continue
+			}
+
+			// Calculate the round-trip time
+			end := time.Now()
+			rtt := end.Sub(entry.LastSendTime)
+
+			// Update the latency store
+			m.latencyStore.UpdateNodeIPLatencyEntryByKey(destIP, &NodeIPLatencyEntry{
+				SeqID:           entry.SeqID,
+				LastSendTime:    entry.LastSendTime,
+				LastRecvTime:    end,
+				LastMeasuredRTT: rtt,
+			})
+		}
+	}
+}
+
+func (m *MonitorTool) PingAll(stopCh <-chan struct{}) {
+	// Create a new socket for IPv4
+	ipv4Socket, err := icmp.ListenPacket(IPv4ProtocolICMPRaw, "0.0.0.0")
+	if err != nil {
+		klog.ErrorS(err, "Failed to create ICMP socket for IPv4")
+		return
+	}
+	defer ipv4Socket.Close()
+	// Create a new socket for IPv6
+	ipv6Socket, err := icmp.ListenPacket(IPv6ProtocolICMPRaw, "::")
+	if err != nil {
+		klog.ErrorS(err, "Failed to create ICMP socket for IPv6")
+		return
+	}
+	defer ipv6Socket.Close()
+
+	// Start the goroutine to receive ICMP messages
+	go m.recvPing(ipv4Socket, true, stopCh)
+	go m.recvPing(ipv6Socket, false, stopCh)
+
+	// Start to ping all nodes
+	pingAll := func() {
+		m.pingAll(ipv4Socket, ipv6Socket)
+	}
+	go wait.Until(pingAll, m.latencyConfig.Interval, stopCh)
```

Comment:
There is a bit too much "nesting" here in my opinion, and it's hard to keep track of what's going on. 

Maybe we could start with a ticker (https://pkg.go.dev/time#Ticker) in the main loop (`monitorLoop`) to send the pings. When the config is changed, we adjust the ticker. For the receiving goroutines, we never need to restart them even if the config is changed.
cc @Dyanngg

Reply:
So I need to put the pingAll logic into monitorLoop, while only needing Ticker time when updating, and stopping these goroutines directly when receiving the signal to delete the CRD.